src/server/game/Entities/Pet/Pet.cpp

```diff
@@ -135,18 +144,67 @@ bool Pet::LoadPetFromDB(Player* owner, uint32 petEntry, uint32 petnumber, bool c
         stmt->setUInt8(2, uint8(PET_SAVE_LAST_STABLE_SLOT));
     }
 
-    result = CharacterDatabase.Query(stmt);
+    WorldSession* session = owner->GetSession();
+    Pet* pet = this;
+    session->GetQueryProcessor().AddQuery(CharacterDatabase.AsyncQuery(stmt)
+        .WithChainingPreparedCallback([pet, session, callback, current](QueryCallback& queryCallback, PreparedQueryResult result)
+    {
+        Player* player = session->GetPlayer();
+        if (!player || !player->IsInWorld())
+        {
+            delete pet;
```

Comment:
What exactly makes "delete this" dangerous if you return immediately after it? I've already seen this pattern in the core. Is there any alternate solution?